### ISSUE TITLE
filesystem bindings

### ISSUE THREAD

**Issue by mitchtbaum** [foundpatterns/torchbear#1]
improvements to [filesystem bindings](https://github.com/foundpatterns/torchbear/blob/master/src/bindings/system/fs.rs), from [lighttouch's fs workarounds](https://github.com/foundpatterns/lighttouch-base/blob/master/fs.lua) and [gut's fs workarounds](https://github.com/foundpatterns/gut/blob/master/extras.lua)

- [x] [rm -r](https://doc.rust-lang.org/std/fs/fn.remove_dir_all.html)
- [x] [touch](https://github.com/foundpatterns/torchbear/commit/c2f9fc6eef2c29c8520f8b00300981f750d90736#diff-adbcdfda7e8740bf97ef918f6190cfa2)
- [x] [join](https://github.com/foundpatterns/torchbear/commit/47024c60f3334a683f33f547cdc32c2cbd6d7e74#diff-3428af05ba62ebcd4ebc826e08011e2f)
- [x] [symlink](https://doc.rust-lang.org/std/fs/index.html?search=symlink)
- [x] current working directory
- [ ] remove_file
- [ ] move_file
- [ ] move_dir

**Comment by mitchtbaum** [comment]
read file
...?

**Comment by Arnaz87** [comment]
read file is in the lua standard io library

it's not crazy to have them separated, some languages have separate libraries for io (read/write file content) and filesystem (directories and metadata), rust being one

**Comment by mitchtbaum** [comment]
append to start of file (prepend)
copy

**Comment by mitchtbaum** [comment]
[open](https://doc.rust-lang.org/std/fs/struct.File.html#method.open)

**Comment by mitchtbaum** [comment]
https://github.com/foundpatterns/lighttouch-base/blob/master/fs.lua
https://github.com/foundpatterns/gut/blob/master/extras.lua

**Comment by mitchtbaum** [comment]
basename
https://github.com/foundpatterns/machu-picchu/pull/9/files#diff-1f2dfa567dcf95833eddf7aec167fec7R17

**Comment by sineemore** [comment]
Need something like `fs.chdir` to change current working directory with Lua.

Example usage:
```
local err = fs.chdir("/this/path/exists")
assert(err == nil, "Failed to chdir to existing directory")

local err = fs.chdir("/no/such/path")
assert(err != nil, "Chdir to nonexistent directory succeeded")
```

`err` should be string describing the failed syscall.

**Comment by dariusc93** [comment]
I dont believe rust have anything builtin to change the directory, but using libc, you could call `libc::chdir` but probably would need to create a wrapper around it since such call would be unsafe.

The current set of fs bindings have an unnatural feel and use due to lua bindings not having any direct access anything such as `std::fs::File`. While making use of `fs::read` and `fs::write` can be useful, improving the overall use of the fs bindings can help with functionality with little bottleneck. This can include making use of `std::fs::File` or even `std::fs::OpenOptions` (probably might be the preferred choice for a little more flexibility since they both will return `std::fs::File` anyway but `OpenOptions` expose more options), being able to read/write data seamlessly, be able to seek through the data and call any related functions in a safe manner (in terms of memory and security), which in theory can replace the io calls within lua (especially since rlua doesnt exactly provide safety with embedded lua it has so having alternatives would be best in this case).

**Comment by sineemore** [comment]
https://doc.rust-lang.org/std/env/fn.set_current_dir.html

**Comment by sineemore** [comment]
> While making use of fs::read and fs::write can be useful.

There is `file:read("a")` in Lua to read all file directly to memory from the start or current position of file handle.

**Comment by dariusc93** [comment]
Yea I saw that function not to long ago (I dont really look in std::env) so I can implement it in a few.

Yea, lua does have that but the problem is that such calls isnt exactly done in a safe manner and rlua doesnt exactly provide any safety for many (if not all) of lua internal functions.

**Comment by sineemore** [comment]
@dariusc93
Oh, where can I read more about that?

**Comment by dariusc93** [comment]
Alot of the information is in rlua description. Other parts can be done by reviewing rlua code and how it interact with the embedded lua code, though everything there is considered to be WIP so things will likely change (which may or may not be breaking on rust side).

**Comment by mitchtbaum** [comment]
https://github.com/kyren/rlua/issues/38#issuecomment-325875248

**Comment by mitchtbaum** [comment]
get current dir
https://doc.rust-lang.org/std/env/fn.current_dir.html

**Comment by dariusc93** [comment]
I have this already in a local branch. Let me rebase and ill push it

**Comment by mitchtbaum** [comment]
is writeable

**Comment by mitchtbaum** [comment]
```lua
#!/usr/bin/env torchbear

print(cwd)
```
returns `nil`

I think that's holding back https://github.com/foundpatterns/torchbear-libs/issues/7

**Comment by dariusc93** [comment]
There is no `cwd` in torchbear. You can use `fs.current_dir()` to get the current directory.

**Comment by mitchtbaum** [comment]
cool.  thanks

**Comment by mitchtbaum** [comment]
reopening for https://github.com/foundpatterns/mp-installer/blob/9ee292956dcec2775068707f2fa8c1e91f4731e5/init.lua#L46-L48
